pkg/kubelet/cm/deviceplugin/manager.go

```diff
@@ -128,20 +132,27 @@ func newManagerImpl(socketPath string) (*ManagerImpl, error) {
 func (m *ManagerImpl) genericDeviceUpdateCallback(resourceName string, added, updated, deleted []pluginapi.Device) {
 	kept := append(updated, added...)
 	m.mutex.Lock()
-	if _, ok := m.allDevices[resourceName]; !ok {
-		m.allDevices[resourceName] = sets.NewString()
+	if _, ok := m.healthyDevices[resourceName]; !ok {
+		m.healthyDevices[resourceName] = sets.NewString()
+	}
+	if _, ok := m.unhealthyDevices[resourceName]; !ok {
+		m.unhealthyDevices[resourceName] = sets.NewString()
 	}
-	// For now, Manager only keeps track of healthy devices.
-	// TODO: adds support to track unhealthy devices.
 	for _, dev := range kept {
 		if dev.Health == pluginapi.Healthy {
-			m.allDevices[resourceName].Insert(dev.ID)
+			m.healthyDevices[resourceName].Insert(dev.ID)
+			m.unhealthyDevices[resourceName].Delete(dev.ID)
 		} else {
-			m.allDevices[resourceName].Delete(dev.ID)
+			m.unhealthyDevices[resourceName].Insert(dev.ID)
+			m.healthyDevices[resourceName].Delete(dev.ID)
 		}
 	}
 	for _, dev := range deleted {
-		m.allDevices[resourceName].Delete(dev.ID)
+		if dev.Health == pluginapi.Healthy {
```

Comment:
The dev health state reported here may not be consistent with the cached state. Maybe simply do:
m.healthyDevices[resourceName].Delete(dev.ID)
m.unhealthyDevices[resourceName].Delete(dev.ID)